### ISSUE TITLE
DbSet<T>.OfType<TDerived>() does not work

### ISSUE THREAD

**Issue by henrikdahl8240** [dotnet/efcore#20056]
I am having constructions like:
`var TDerivedInstance = await myContext.DbSetOfT.OfType<TDerived>().FirstAsync(x=> ... some query ...);`

Unfortunately it gives an exception:
```
System.InvalidOperationException: The LINQ expression 'DbSet<T>
    .OfType()' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to either AsEnumerable(), AsAsyncEnumerable(), ToList(), or ToListAsync(). See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
```
At this article, it actually reads, that OfType<TDerived>() should work: https://stackoverflow.com/questions/57940222/why-there-is-no-ef-core-oftypetype-type-method where it reads: "The generic method OfType<T> is a standard LINQ method which EF Core supports.

There is no standard OfType(Type) method and EF Core designers didn't find a reason of implementing such custom EF Core specific extension.".

Am I missing something since OfType<TDerived>() doesn't work for me?

I think that it's worth to notice, that OfType<TDerived>() returns an iQueryable<TDerived> and therefore it does not compare to DbSet<T>.where(x => x is TDerived) because it returns an IQueryable<T> and not IQueryable<TDerived>.

Basically, how to achieve the funtionality of DbSet<T>.OfType<TDerived>() from EF 6 using EF Core 3.1.2?

EF Core version: 3.1.2.
Database provider: Microsoft.EntityFrameworkCore.SqlServer.
Target framework: (e.g. .NET Core 3.1.2)
Operating system: Windows 10 Professional.
IDE: Visual Studio 2019 16.4.5

**Comment by henrikdahl8240** [comment]
In my post I used the generic variant of OfType(), but it doesn't appear with the generic argument specified, perhaps due to some conflict with markdown.

**Comment by henrikdahl8240** [comment]
... or should DbContext.DbSet<T>.OfType<TDerived>() be replaced by DbContext.Set<TDerived>()?

**Comment by smitpatel** [comment]
`DbSet<T>().OfType<TDerived>()` works as long as TDerived is also part of the model. We have existing tests verifying that.
https://github.com/dotnet/efcore/blob/ce8431c63d978852beda1e938b0e1caf048af466/test/EFCore.Specification.Tests/Query/InheritanceTestBase.cs#L114-L124

Is TDerived you are using registered as entityType in the model? If yes then can you share a runnable repro code which we can investigate.

**Comment by henrikdahl8240** [comment]
Unfortunately I have realized, that a few abstract classes, including TDerived, were not registered as an entityType in the model. After I have registered these types, OfType<TDerived>() works correctly.

I would like to excuse, that I indicated, that there was an error in Entity Framework Core. I may conclude, that this case did not show that.

Perhaps it could be an idea to include this hint in the error message in addition to the other focus points I quoted in my initial mail.
